### PR TITLE
[Relay] [Op] add relu

### PR DESCRIPTION
**Pull request by MarisaKirisame**
I had lifted RELAY_REGISTER_UNARY_OP as I found relu also needing it.

### CODE REVIEW

File: include/tvm/relay/op.h [hunk 1]
```diff
@@ -316,6 +316,52 @@ class OpMap {
...
+ *
+ * We make the decision to always only expose positional argument.
+ * We will do rewrapping in the frontend to support language
+ * sugars such as keyword arguments and default value.
+ *
+ * \param Prefix the prefix of the registry, for example, "relay.op._make.".
+ *
+ * \param OpName the name of registry.
+ */
+#define RELAY_REGISTER_UNARY_OP(Prefix, OpName)           \
```
Comment (tqchen) [inline include/tvm/relay/op.h]: put these helpers in src/relay/op/op_common.h

File: python/tvm/relay/op/nn/nn.py [hunk 2]
```diff
@@ -86,3 +86,19 @@ def conv2d(data,
     return _make.conv2d(data, weight, strides, padding, dilation,
                         groups, channels, kernel_size, data_layout,
                         weight_layout, out_layout, out_dtype)
+
+
+def relu(data):
+    """Compute elementwise relu of data.
```
Comment (tqchen) [inline python/tvm/relay/op/nn/nn.py]: Rectified linear unit, add an math expression about relu.

File: src/relay/op/nn/relu.cc [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+#include "../type_relations.h"
+
+namespace tvm {
+namespace relay {
+
+RELAY_REGISTER_UNARY_OP("relay.op.nn._make.", "relu")
+.describe(R"code(Returns the relu input array, computed element-wise.
+
+.. math::
+   relu(x)
```
Comment (tqchen) [inline src/relay/op/nn/relu.cc]: max(x, 0)

File: src/relay/op/tensor/binary.cc [hunk 4]
```diff
@@ -10,46 +10,36 @@
...
-  RELAY_REGISTER_OP(OpName)                                            \
-  .set_num_inputs(2)                                                   \
-  .add_argument("lhs", "Tensor", "The left hand side tensor.")         \
-  .add_argument("rhs", "Tensor", "The right hand side tensor.")        \
-  .add_type_rel("Broadcast", BroadcastRel)
+#define RELAY_MAKE_REGISTER_BINARY_OP(OpName) RELAY_REGISTER_BINARY_OP("relay.op._make.", OpName)
 
 // Addition
-RELAY_REGISTER_BINARY_OP("add")
+RELAY_MAKE_REGISTER_BINARY_OP("add")
```
Comment (tqchen) [inline src/relay/op/tensor/binary.cc]: keep the original name and marcro

Comment (MarisaKirisame) [reply src/relay/op/tensor/binary.cc]: what should be the name of RELAY_REGISTER_UNARY_OP and RELAY_MAKE_REGISTER_BINARY_OP? I dont think having them the same is a good idea

Comment (tqchen) [reply src/relay/op/tensor/binary.cc]: I thought both name ae RELAY_REGISTER_UNARY_OP AND RELAY_REGISTER_BINARY_OP?

Comment (MarisaKirisame) [reply src/relay/op/tensor/binary.cc]: Yes, but what should be the name of the macro in op_common? or should we just call it directly (without helper macro that wrap it)?

**Comment by tqchen** [comment]
I see what you mean, given that it is not hard to directly expand the macro for relu, let us simply do that here. It is a bit confusing to have nested macro for not so common case

File: src/relay/op/tensor/unary.cc [hunk 5]
```diff
@@ -153,6 +137,18 @@ RELAY_REGISTER_UNARY_OP("negative")
 .set_support_level(3)
 .add_type_rel("Identity", IdentityRel);
 
+// Concat
+TVM_REGISTER_API("relay.op._make.concat")
+  .set_body_typed<Expr(Expr)>([](Expr tuple) {
+    static const Op& op = Op::Get("concat");
+    return CallNode::make(op, { tuple }, Attrs(), {});
+  });
+
```
Comment (tqchen) [inline src/relay/op/tensor/unary.cc]: concat should not be part of this PR

File: src/relay/op/nn/relu.cc [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#include <tvm/relay/expr.h>
+#include <tvm/relay/op.h>
+#include "../type_relations.h"
+#include "../op_common.h"
+
+namespace tvm {
+namespace relay {
+
+RELAY_REGISTER_UNARY_OP("relay.op.nn._make.", "relu")
+.describe(R"code(Returns the relu input array, computed element-wise.
```
Comment (tqchen) [inline src/relay/op/nn/relu.cc]: directly put it in nn.cc is fine.

File: docs/langref/relay_op.rst [hunk 7]
```diff
@@ -112,8 +114,10 @@ Level 1 Definitions
 .. autofunction:: tvm.relay.mod
 .. autofunction:: tvm.relay.tanh
 .. autofunction:: tvm.relay.sigmoid
+.. autofunction:: tvm.relay.expand_dims
```
Comment (tqchen) [inline docs/langref/relay_op.rst]: expand_dims do not belong to here
